File: src/domain/notifications.ts
import { formatMoney } from './formatMoney'
import type { BudgetStatus } from './budget'
import type { Id, IsoDate } from './entities/types'

/**
 * الإشعارات — دوال نقية تبني التنبيهات ولا ترسل حاجة.
 *
 * ⚠️ **حد تقني حقيقي، مش اختصار:** الدفع من السيرفر (server push)
 * مستحيل هنا. الإشعار اللي بيوصل الموبايل والتطبيق مقفول محتاج
 * Cloud Functions، وهي مش متاحة في باقة Spark (CLAUDE.md #12).
 * فاللي بيتبنى هنا **تنبيهات داخل التطبيق** بتظهر لما يفتحه.
 * ده مكتوب في الشاشة نفسها كمان، عشان محدش يستنى تنبيه مش هييجي.
 *
 * القاعدة الحاكمة (`spec/06`): «تصنيف بلا سقف/تاريخ ⇒ **لا سقف أو
 * متوسط مخترع ولا تنبيه عتبة**». يعني التنبيه ما بيتولدش إلا من رقم
 * حطه المستخدم بنفسه.
 */

export type NotificationKind = 'budget_total' | 'budget_category' | 'recurring_due'

export type NotificationSeverity = 'info' | 'warn' | 'over'

export interface NotificationEvent {
  /**
   * مفتاح الحدث — بيمنع تكرار نفس التنبيه.
   *
   * بيضم الفترة والهدف والعتبة، فعبور ٨٠٪ بينبّه مرة، وعبور ١٠٠٪
   * بعدها بينبّه مرة تانية لأنه **حدث تاني** مش تكرار للأول.
   */
  eventKey: string
  kind: NotificationKind
  severity: NotificationSeverity
  title: string
  body: string
  periodStart: IsoDate
  categoryId?: Id
  recurringId?: Id
  /** العتبة اللي اتعدّت، بالمئة. */
  threshold?: number
}

/**
 * إيصال تنبيه اتبعت — `spec/03`:
 * «NotificationReceipt: eventKey، threshold، category/recurringId،
 *  periodStart، sentAt؛ **لمنع تكرار التنبيه**»
 */
export interface NotificationReceipt {
  eventKey: string
  threshold: number | null
  categoryId?: Id
  recurringId?: Id
  periodStart: IsoDate
  sentAt: string
}

/** العتبات المعلنة. عتبة المستخدم بتتضاف لهم لو مختلفة. */
const HARD_THRESHOLDS = [100] as const

function severityFor(threshold: number): NotificationSeverity {
  if (threshold >= 100) return 'over'
  return threshold >= 90 ? 'warn' : 'info'
}

/**
 * العتبات اللي عدّاها سقف واحد فعلًا، مرتبة.
 *
 * بترجّع **كل** عتبة اتعدّت لا الأعلى بس، عشان لو المستخدم مافتحش
 * التطبيق أيام ما يضيعش تنبيه ٨٠٪ لمجرد إنه بقى فوق ١٠٠٪.
 */
function crossedThresholds(status: BudgetStatus, userThreshold: number | null): number[] {
  const candidates = new Set<number>(HARD_THRESHOLDS)
  // عتبة المستخدم هي الوحيدة اللي هو طلبها — بتتحط جنب عتبة التجاوز
  if (userThreshold !== null && userThreshold > 0) candidates.add(userThreshold)

  return [...candidates]
    .filter((threshold) => status.usedTenthPercent >= threshold * 10)
    .sort((a, b) => a - b)
}

export interface BudgetNotificationInput {
  periodStart: IsoDate
  /** حالة السقف الإجمالي، أو null لو مفيش سقف إجمالي. */
  totalStatus: BudgetStatus | null
  /** العتبة اللي حددها المستخدم للسقف الإجمالي. */
  totalThresholdPercent: number | null
  /**
   * التصنيفات اللي **ليها سقف**. اللي بلا سقف ما بيتبعتش أصلًا،
   * فمفيش احتمال تنبيه عتبة على تصنيف بلا سقف.
   */
  categories: readonly {
    categoryId: Id
    categoryName: string
    status: BudgetStatus
    thresholdPercent: number | null
  }[]
  /**
   * هل المصروف معروف أصلًا؟ `false` لما تكون فيه عمليات بلا نوع
   * اقتصادي محدد — وقتها الرقم ناقص، والتنبيه عليه تخويف بلا أساس.
   */
  spentKnown: boolean
}

/** يبني تنبيهات الميزانية. بيحسب مش بيقرر إمتى تتعرض. */
export function buildBudgetNotifications(
  input: BudgetNotificationInput,
): NotificationEvent[] {
  // رقم ناقص ما ينفعش يتبنى عليه تنبيه (CLAUDE.md #10)
  if (!input.spentKnown) return []

  const events: NotificationEvent[] = []

  if (input.totalStatus) {
    for (const threshold of crossedThresholds(input.totalStatus, input.totalThresholdPercent)) {
      events.push({
        eventKey: `${input.periodStart}|total|${threshold}`,
        kind: 'budget_total',
        severity: severityFor(threshold),
        threshold,
        periodStart: input.periodStart,
        title: threshold >= 100 ? 'عدّيت الميزانية' : `وصلت ${threshold}٪ من الميزانية`,
        body: describeStatus(input.totalStatus),
      })
    }
  }

  for (const category of input.categories) {
    for (const threshold of crossedThresholds(category.status, category.thresholdPercent)) {
      events.push({
        eventKey: `${input.periodStart}|cat:${category.categoryId}|${threshold}`,
        kind: 'budget_category',
        severity: severityFor(threshold),
        threshold,
        categoryId: category.categoryId,
        periodStart: input.periodStart,
        title:
          threshold >= 100
            ? `عدّيت سقف «${category.categoryName}»`
            : `«${category.categoryName}» وصل ${threshold}٪`,
        body: describeStatus(category.status),
      })
    }
  }

  // الأخطر الأول، ثم الأعلى عتبة
  return events.sort((a, b) => (b.threshold ?? 0) - (a.threshold ?? 0))
}

/** جملة الحالة بالريال — الهللة وحدة داخلية لا تظهر للمستخدم أبدًا. */
function describeStatus(status: BudgetStatus): string {
  const spent = formatMoney(status.spentMinor)
  const limit = formatMoney(status.limitMinor)
  if (status.remainingMinor < 0) {
    return `صرفت ${spent} من ${limit} — زيادة ${formatMoney(-status.remainingMinor)}`
  }
  return `صرفت ${spent} من ${limit} — فاضل ${formatMoney(status.remainingMinor)}`
}

/**
 * يشيل اللي اتنبّه عليه قبل كده.
 *
 * ده الغرض الوحيد من `NotificationReceipt`: التنبيه يظهر **مرة واحدة**
 * لكل حدث، فما يتحوّلش لضوضاء يتجاهلها المستخدم.
 */
export function filterUnseen(
  events: readonly NotificationEvent[],
  receipts: readonly NotificationReceipt[],
): NotificationEvent[] {
  const seen = new Set(receipts.map((r) => r.eventKey))
  return events.filter((event) => !seen.has(event.eventKey))
}

/** يبني إيصالًا من حدث — نفس المفتاح، فالمنع مضمون. */
export function receiptFor(event: NotificationEvent, sentAt: string): NotificationReceipt {
  return {
    eventKey: event.eventKey,
    threshold: event.threshold ?? null,
    ...(event.categoryId ? { categoryId: event.categoryId } : {}),
    ...(event.recurringId ? { recurringId: event.recurringId } : {}),
    periodStart: event.periodStart,
    sentAt,
  }
}

/**
 * إيصالات الفترات القديمة اللي ما بقاش ليها لازمة.
 * الإيصال بيمنع التكرار داخل فترته؛ بعد ما تعدّي بقى حِمل بلا فايدة.
 */
export function staleReceipts(
  receipts: readonly NotificationReceipt[],
  keepFromPeriodStart: IsoDate,
): NotificationReceipt[] {
  return receipts.filter((r) => r.periodStart < keepFromPeriodStart)
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/Notifications.kt
package app.masroufy.core

/**
 * التنبيهات جوه التطبيق + قفل البصمة + الشعارات — نقل `notifications.ts` + `appLock.ts` + `merchantLogo.ts`.
 * التنبيه **ما بيتولدش إلا من رقم حطه المستخدم** (spec/06)، وما بيتكررش لنفس الحدث (الإيصال).
 */
data class NotificationEvent(
    /** الفترة + الهدف + العتبة — عبور ٨٠٪ حدث، وعبور ١٠٠٪ بعده حدث تاني. */
    val eventKey: String,
    /** "budget_total" / "budget_category" / "recurring_due". */
    val kind: String,
    /** "info" / "warn" / "over". */
    val severity: String,
    val title: String,
    val body: String,
    val periodStart: IsoDate,
    val categoryId: Id? = null,
    val recurringId: Id? = null,
    val threshold: Int? = null,
)

data class NotificationReceipt(
    val eventKey: String,
    val threshold: Int?,
    val periodStart: IsoDate,
    val sentAt: String,
    val categoryId: Id? = null,
    val recurringId: Id? = null,
)

private val HARD_THRESHOLDS = listOf(100)

private fun severityFor(threshold: Int) = when {
    threshold >= 100 -> "over"
    threshold >= 90 -> "warn"
    else -> "info"
}

/** **كل** عتبة اتعدّت، مش الأعلى بس — عشان تنبيه ٨٠٪ ما يضيعش لو المستخدم ما فتحش أيام. */
private fun crossedThresholds(status: BudgetStatus, userThreshold: Int?): List<Int> {
    val candidates = LinkedHashSet(HARD_THRESHOLDS)
    if (userThreshold != null && userThreshold > 0) candidates += userThreshold
    return candidates.filter { status.usedTenthPercent >= it * 10L }.sorted()
}

data class CategoryBudgetForNotice(val categoryId: Id, val categoryName: String, val status: BudgetStatus, val thresholdPercent: Int?)

private fun describeStatus(status: BudgetStatus): String {
    val spent = formatMoney(status.spentMinor)
    val limit = formatMoney(status.limitMinor)
    return if (status.remainingMinor < 0) "صرفت $spent من $limit — زيادة ${formatMoney(-status.remainingMinor)}"
    else "صرفت $spent من $limit — فاضل ${formatMoney(status.remainingMinor)}"
}

/** تنبيهات الميزانية. المصروف مش معروف (عمليات من غير نوع) ⇒ مفيش تنبيه على رقم ناقص (CLAUDE.md #10). */
fun buildBudgetNotifications(
    periodStart: IsoDate,
    totalStatus: BudgetStatus?,
    totalThresholdPercent: Int?,
    categories: List<CategoryBudgetForNotice>,
    spentKnown: Boolean,
): List<NotificationEvent> {
    if (!spentKnown) return emptyList()
    val events = mutableListOf<NotificationEvent>()
    if (totalStatus != null) {
        for (t in crossedThresholds(totalStatus, totalThresholdPercent)) {
            events += NotificationEvent(
                "$periodStart|total|$t", "budget_total", severityFor(t),
                if (t >= 100) "عدّيت الميزانية" else "وصلت $t٪ من الميزانية", describeStatus(totalStatus), periodStart, threshold = t,
            )
        }
    }
    for (c in categories) {
        for (t in crossedThresholds(c.status, c.thresholdPercent)) {
            events += NotificationEvent(
                "$periodStart|cat:${c.categoryId}|$t", "budget_category", severityFor(t),
                if (t >= 100) "عدّيت سقف «${c.categoryName}»" else "«${c.categoryName}» وصل $t٪", describeStatus(c.status), periodStart,
                categoryId = c.categoryId, threshold = t,
            )
        }
    }
    return events.sortedByDescending { it.threshold ?: 0 }
}

fun filterUnseen(events: List<NotificationEvent>, receipts: List<NotificationReceipt>): List<NotificationEvent> {
    val seen = receipts.map { it.eventKey }.toSet()
    return events.filter { it.eventKey !in seen }
}

fun receiptFor(event: NotificationEvent, sentAt: String) = NotificationReceipt(
    event.eventKey, event.threshold, event.periodStart, sentAt,
    event.categoryId?.ifEmpty { null }, event.recurringId?.ifEmpty { null },
)

/** إيصالات الفترات اللي عدّت — حِمل من غير فايدة. */
fun staleReceipts(receipts: List<NotificationReceipt>, keepFromPeriodStart: IsoDate): List<NotificationReceipt> =
    receipts.filter { it.periodStart < keepFromPeriodStart }

/* ───────────────────────── قفل البصمة (OVERRIDES §21) ───────────────────────── */

const val LOCK_AFTER_BACKGROUND_MS = 5 * 60_000L

/** مقفول لحد ما يتشغل؛ أول فتح بيقفل؛ ٥ دقايق في الخلفية بيقفل؛ الساعة رجعت لورا ⇒ بيقفل احتياطًا. */
fun shouldLock(enabled: Boolean, hiddenAt: Long?, now: Long): Boolean {
    if (!enabled) return false
    if (hiddenAt == null) return true
    if (now < hiddenAt) return true
    return now - hiddenAt >= LOCK_AFTER_BACKGROUND_MS
}

/* ───────────────────────── شعارات التجار (OVERRIDES §25.1) ───────────────────────── */

data class MerchantLogoEntry(val names: List<String>, val file: String? = null, val domain: String? = null)

sealed interface LogoSource {
    data class Bundled(val file: String) : LogoSource
    data class Online(val domain: String) : LogoSource
}

private val DOMAIN = Regex("^[a-z0-9]([a-z0-9-]*[a-z0-9])?(\\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)+$")

/** أول تعريف للاسم هو اللي بيفضل. */
fun buildLogoIndex(entries: List<MerchantLogoEntry>): Map<String, MerchantLogoEntry> {
    val index = LinkedHashMap<String, MerchantLogoEntry>()
    for (e in entries) for (name in e.names) {
        val key = normalizeText(name)
        if (key.isNotEmpty() && key !in index) index[key] = e
    }
    return index
}

/** الملف اللي جوه التطبيق بيغلب دايمًا؛ الأونلاين بس لو اتفعّل والدومين سليم. */
fun logoSourceFor(merchantName: String?, index: Map<String, MerchantLogoEntry>, onlineEnabled: Boolean, bundledFiles: Set<String>): LogoSource? {
    val entry = index[normalizeText(merchantName ?: "")] ?: return null
    if (!entry.file.isNullOrEmpty() && entry.file in bundledFiles) return LogoSource.Bundled(entry.file)
    val domain = entry.domain?.let(JsText::trim)?.lowercase()
    if (onlineEnabled && !domain.isNullOrEmpty() && DOMAIN.matches(domain)) return LogoSource.Online(domain)
    return null
}
